File: algorithms/sorting/radix.js
// =====================================================================
// Radix Sort (LSD) / Поразрядная сортировка
// ---------------------------------------------------------------------
// Идея: сортируем по разрядам от младшего к старшему (LSD = least
//       significant digit). На каждом шаге применяем стабильную
//       сортировку по текущему разряду (обычно counting sort).
// Time:  O(d · (n + k)), d = количество разрядов, k = база (10 здесь)
// Space: O(n + k)
// Stable: yes
// In-place: no
// Условие: неотрицательные целые. Для отрицательных нужна модификация.
// =====================================================================

function radixSort(arr) {
    if (arr.length === 0) return arr;

    const max = Math.max(...arr);
    // exp — текущий разряд: 1 (единицы), 10 (десятки), 100 ...
    for (let exp = 1; Math.floor(max / exp) > 0; exp *= 10) {
        countingByDigit(arr, exp);
    }

    return arr;
}

/**
 * Counting sort, который сортирует по конкретному разряду exp.
 * Извлекаем разряд через (num / exp) % 10
 */
function countingByDigit(arr, exp) {
    const n = arr.length;
    const output = new Array(n);
    const count = new Array(10).fill(0); // десятичная база

    // Подсчёт цифр на разряде exp
    for (let i = 0; i < n; i++) {
        const digit = Math.floor(arr[i] / exp) % 10;
        count[digit]++;
    }

    // Префиксные суммы
    for (let i = 1; i < 10; i++) {
        count[i] += count[i - 1];
    }

    // Восстановление с конца (стабильность критична!)
    for (let i = n - 1; i >= 0; i--) {
        const digit = Math.floor(arr[i] / exp) % 10;
        count[digit]--;
        output[count[digit]] = arr[i];
    }

    // Копируем обратно в arr
    for (let i = 0; i < n; i++) arr[i] = output[i];
}

console.log(radixSort([170, 45, 75, 90, 802, 24, 2, 66])); // [2, 24, 45, 66, 75, 90, 170, 802]

module.exports = radixSort;

File: algorithms/sorting/radix.py
"""
=====================================================================
Radix Sort (LSD) / Поразрядная сортировка
---------------------------------------------------------------------
Идея: сортируем по разрядам от младшего к старшему. На каждом разряде —
      стабильная сортировка (counting sort).
Time:  O(d · (n + k)), d = разряды, k = база
Space: O(n + k)
Stable: yes
In-place: no
Условие: неотрицательные целые.
=====================================================================
"""
from typing import List


def radix_sort(arr: List[int]) -> List[int]:
    if not arr:
        return arr

    max_num = max(arr)
    exp = 1
    while max_num // exp > 0:
        _counting_by_digit(arr, exp)
        exp *= 10

    return arr


def _counting_by_digit(arr: List[int], exp: int) -> None:
    n = len(arr)
    output = [0] * n
    count = [0] * 10

    for num in arr:
        digit = (num // exp) % 10
        count[digit] += 1

    for i in range(1, 10):
        count[i] += count[i - 1]

    # С конца для стабильности
    for i in range(n - 1, -1, -1):
        digit = (arr[i] // exp) % 10
        count[digit] -= 1
        output[count[digit]] = arr[i]

    for i in range(n):
        arr[i] = output[i]


if __name__ == "__main__":
    print(radix_sort([170, 45, 75, 90, 802, 24, 2, 66]))
    # [2, 24, 45, 66, 75, 90, 170, 802]
